Lib/idlelib/autocomplete.py

```diff
@@ -90,6 +92,126 @@ def _delayed_open_completions(self, args):
+                return None
+            comp_start = m.group(1) + hp.rawtext[opening_quote_idx:hp.indexinrawtext]
+        elif hp.is_in_code():
+            if not hp.isopener[hp.indexbracket]:
+                return None
+            opening_bracket_idx = hp.bracketing[hp.indexbracket][0]
+            if hp.rawtext[opening_bracket_idx] != '[':
+                return None
+            m = cls._dict_str_key_prefix_re.fullmatch(
+                hp.rawtext[opening_bracket_idx + 1:hp.indexinrawtext])
+            if m is None:
+                return None
+            comp_start = m.group(1)
+        else:
+            return None
+
+        hp.set_index("insert-%dc" % (len(hp.rawtext) - opening_bracket_idx))
+        comp_what = hp.get_expression()
+        return comp_what, comp_start
+
+    @classmethod
+    def dict_key_reprs(cls, comp_start, comp_list):
+        """Customized repr() for str and bytes.
+        
+        This retains the prefix and quote type of the completed partial
+        str/bytes literal, if they exist.
+        """
+        # Analyze existing prefix and quote type of the completion start.
+        is_raw = False
+        quote_type = '"'
+        prefix_match = cls._dict_str_key_prefix_re.match(comp_start)
+        if prefix_match is not None:
+            prefix = prefix_match.group(1)
+            post_prefix = comp_start[prefix_match.end():]
+            lowercase_prefix = prefix.lower()
+
+            is_raw = 'r' in lowercase_prefix
+            if post_prefix:
+                quote_type_candidate = post_prefix[0]
+                if quote_type_candidate not in "'\"":
+                    pass  # use the default quote type
+                elif post_prefix[1:2] == quote_type_candidate:
+                    quote_type = quote_type_candidate * 3
+                else:
+                    quote_type = quote_type_candidate
+        else:
+            prefix = lowercase_prefix = ''
+
+        # Prepare the prefixes to use for the reprs.
+        prefix = prefix.replace('b', '').replace('B', '')
+        bytes_prefix = (prefix + 'b') if lowercase_prefix in ('', 'r') else 'b'
+
+        # Prepare escaping utilities.
+        escape_re = re.compile(r'[\0-\15]')
+        escapes = {chr(x): f'\\{x}' for x in range(7)}
+        escapes.update({
+            7: '\\a',
+            8: '\\b',
+            9: '\\t',
+            10: '\\n',
+            11: '\\v',
+            12: '\\f',
+            13: '\\r',
+        })
+        high_byte_escape_re = re.compile(r'[\x80-\xff]')
+        def high_byte_sub(char):
+            """escaping for bytes values above 127"""
+            return f'\\x{ord(char):x}'
+
+        # Split the completions into lists of str and bytes.
+        str_comps = (c for c in comp_list if type(c) == str)
+        bytes_comps = (c for c in comp_list if type(c) == bytes)
+
+        # Create reprs for the str objects.
+        str_repr_template = f"{prefix}{quote_type}{{}}{quote_type}"
```

Comment:
Minor QoL suggestion:

In any scenario I can imagine for the dict key auto-complete, the user would end up typing `]` immediately after using `tab`. As a result, it might be beneficial to add it to the end of the auto-completion.

As far as I can tell, the best place to add this would be in the template:
```suggestion
        str_repr_template = f"{prefix}{quote_type}{{}}{quote_type}]"
```

Reply:
Tested this suggestion, no issues as far as I can tell.

Before tab:
![image](https://user-images.githubusercontent.com/44193521/62675809-7fa04480-b976-11e9-90fd-9f01bffaf26c.png)
After tab:
![image](https://user-images.githubusercontent.com/44193521/62675824-8c249d00-b976-11e9-891b-682c52d79d7d.png)

Reply:
This is a great idea and I've just implemented it, albeit using a more robust and complete approach.

Reply:
Awesome! I definitely agree with the usage of a variable instead of appending a constant, that allows far more extensibility. I hadn't thought of using anything besides `]`, but that could definitely be useful.